### PR TITLE
2x and 4x oversampling for WaveShaperNode

### PR DESCRIPTION
**Pull request by khodzha**
Issue #205 
Finally managed to defeat the noise I mentioned in the issue 

I'm not sure if ditching FrameIterator and operating on raw vecs was a good thing :man_shrugging: 

Also spec mentions tail-time but I'm not sure how to handle that

### CODE REVIEW

File: audio/Cargo.toml [hunk 1]
```diff
@@ -18,6 +18,10 @@ servo-media-player = { path = "../player" }
 servo-media-traits = { path = "../traits" }
 smallvec = "0.6.1"
 
+[dependencies.speexdsp]
+git = "https://github.com/rust-av/speexdsp-rs"
```
Comment (Manishearth) [inline audio/Cargo.toml]: Not https://crates.io/crates/speexdsp ?

File: audio/wave_shaper_node.rs [hunk 2]
```diff
@@ -116,3 +165,30 @@ impl AudioNodeEngine for WaveShaperNode {
 
     make_message_handler!(WaveShaperNode: handle_waveshaper_message);
 }
+
+impl WaveShaperNode {
+    fn apply_curve(buf: &mut [f32], curve: &Vec<f32>) {
+        buf.iter_mut().for_each(|sample| {
+            let len = curve.len();
+            let curve_index: f32 = ((len - 1) as f32) * (*sample + 1.) / 2.;
```
Comment (Manishearth) [inline audio/wave_shaper_node.rs]: worth  moving `(len - 1) / 2` and `len` out of the closure. Probably will be done by the compiler anyway, but doesn't hurt.

File: audio/Cargo.toml (hunk 1 above)
Comment (Manishearth) [reply audio/Cargo.toml]: Ah, it's not released yet.

Comment (Manishearth) [reply audio/Cargo.toml]: Looks like they're willing to make one soon! https://github.com/rust-av/speexdsp-rs/issues/10

File: audio/wave_shaper_node.rs (hunk 2 above)
Comment (khodzha) [reply audio/wave_shaper_node.rs]: agreed, I'll fix it up tomorrow

**Comment by lu-zero** [comment]
I prepared a crate with the stand alone resampler, if you could please confirm that the API exposed is fine for you I'll publish it.

**Comment by Manishearth** [comment]
@khodzha ^^

**Comment by khodzha** [comment]
Updated PR to use this crate, seems ok to me :man_shrugging:

**Comment by lu-zero** [comment]
https://crates.io/crates/speexdsp-resampler published !

File: audio/wave_shaper_node.rs [hunk 3]
```diff
@@ -79,33 +93,68 @@ impl AudioNodeEngine for WaveShaperNode {
         AudioNodeType::WaveShaperNode
     }
 
-    fn process(&mut self, mut inputs: Chunk, _info: &BlockInfo) -> Chunk {
+    fn process(&mut self, mut inputs: Chunk, info: &BlockInfo) -> Chunk {
         debug_assert!(inputs.len() == 1);
 
         if inputs.blocks[0].is_silence() {
```
Comment (khodzha) [inline audio/wave_shaper_node.rs]: @Manishearth could you confirm whether tail-time here could be implemented like this:
if we hit two silent blocks in row after non-silent block, we `explicit_silence()` them and run them through the waveshaping so all the tail-time from resampler would be spitted out

Comment (Manishearth) [reply audio/wave_shaper_node.rs]: That sounds correct.

**Comment by khodzha** [comment]
Added tail-time handling and properly handled the case when waveshaped silence is nonsilent

**Comment by Manishearth** [comment]
@bors-servo r+

thanks!

**Pull request merged**